### ISSUE TITLE
Bug: allowAutomaticSingleRunInference not working in pnpm repository

### ISSUE THREAD

**Issue by xqxian** [typescript-eslint/typescript-eslint#8593]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Repro Code

```TypeScript
# Run eslint from command line in a pnpm repository
eslint any-file.ts --ext .ts
```

### ESLint Config

```javascript
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    allowAutomaticSingleRunInference: true,
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

typescript-eslint detects single run mode successfully

### Actual Result

typescript-eslint detects single run mode failed

### Additional Info

I have found that the cause of the problem lies in the file path detection for enabling single run mode([source](https://github.com/typescript-eslint/typescript-eslint/blob/c4310b1aac35c7d31b826f0602eca6a5900a09ee/packages/typescript-estree/src/parser.ts#L390)).

When running eslint from the command line in a pnpm repository, the eslint bin file path will end with `node_modules/eslint/bin/eslint.js` so the single run mode detection will fail:

!["Screenshot of running breakpoint"](https://github.com/typescript-eslint/typescript-eslint/assets/21199814/0348d61a-1670-43dd-b2ed-1b2c9aa47487)

 I am currently working on a pnpm monorepo that contains more than 40 TypeScript projects. I would like to use `allowAutomaticSingleRunInference` to enhance performance and would be happy to submit a PR to fix it. I know that by setting the environment variable `TSESTREE_SINGLE_RUN=true`, I can force typescript-eslint to run in single run mode. However, relying solely on this approach may not be optimal as it can be easy to forget to set the environment variable when running eslint manually from the command line.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `7.1.0` |
| `@typescript-eslint/parser`            | `7.1.0` |
| `@typescript-eslint/rule-tester`       | `7.1.0` |
| `@typescript-eslint/scope-manager`     | `7.1.0` |
| `@typescript-eslint/typescript-estree` | `7.1.0` |
| `@typescript-eslint/type-utils`        | `7.1.0` |
| `@typescript-eslint/utils`             | `7.1.0` |
| `TypeScript`                           | `4.9.5` |
| `ESLint`                               | `8.57.0` |
| `node`                                 | `18.18.1` |
|`pnpm`                                | `6.35.1` |

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/3811

Please use the search
